presto-raptor/src/main/java/com/facebook/presto/raptor/storage/BucketBalancer.java

```diff
@@ -0,0 +1,369 @@
+ * produce a cluster state that is imbalanced in terms of disk utilization.
+ */
+public class BucketBalancer
+{
+    private static final Logger log = Logger.get(BucketBalancer.class);
+
+    private final NodeSupplier nodeSupplier;
+    private final ShardManager shardManager;
+    private final boolean enabled;
+    private final Duration interval;
+    private final boolean backupAvailable;
+    private final boolean coordinator;
+    private final ScheduledExecutorService executor;
+
+    private final AtomicBoolean started = new AtomicBoolean();
+
+    private final CounterStat bucketsBalanced = new CounterStat();
+    private final CounterStat jobErrors = new CounterStat();
+
+    @Inject
+    public BucketBalancer(
+            NodeManager nodeManager,
+            NodeSupplier nodeSupplier,
+            ShardManager shardManager,
+            BucketBalancerConfig config,
+            BackupService backupService,
+            RaptorConnectorId connectorId)
+    {
+        this(nodeSupplier,
+                shardManager,
+                config.isBalancerEnabled(),
+                config.getBalancerInterval(),
+                backupService.isBackupAvailable(),
+                nodeManager.getCurrentNode().isCoordinator(),
+                connectorId.toString());
+    }
+
+    public BucketBalancer(
+            NodeSupplier nodeSupplier,
+            ShardManager shardManager,
+            boolean enabled,
+            Duration interval,
+            boolean backupAvailable,
+            boolean coordinator,
+            String connectorId)
+    {
+        this.nodeSupplier = requireNonNull(nodeSupplier, "nodeSupplier is null");
+        this.shardManager = requireNonNull(shardManager, "shardManager is null");
+        this.enabled = enabled;
+        this.interval = requireNonNull(interval, "interval is null");
+        this.backupAvailable = backupAvailable;
+        this.coordinator = coordinator;
+        this.executor = newScheduledThreadPool(1, daemonThreadsNamed("bucket-balancer-" + connectorId));
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (enabled && backupAvailable && coordinator && !started.getAndSet(true)) {
+            startJob();
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        executor.shutdownNow();
+    }
+
+    @Managed
+    @Nested
+    public CounterStat getBucketsBalanced()
+    {
+        return bucketsBalanced;
+    }
+
+    @Managed
+    @Nested
+    public CounterStat getJobErrors()
+    {
+        return jobErrors;
+    }
+
+    private void startJob()
+    {
+        executor.scheduleWithFixedDelay(() -> {
+            try {
+                process();
+            }
+            catch (Throwable t) {
+                log.error(t, "Error balancing distributionAssignments");
+                jobErrors.update(1);
+            }
+        }, 0, interval.toMillis(), MILLISECONDS);
```

Comment:
Running immediately on startup probably isn't a good idea. Maybe we should always wait the full interval for the initial run? Or maybe half the interval?